### PR TITLE
Adding helm chart

### PR DESCRIPTION
**Pull request by cccs-cat001**
I created a helm chart to deploy nbgallery to a kubernetes cluster.

### CODE REVIEW

**Comment by mcrutch** [comment]
I could have missed it (new to helm) but you are going to need some form of persistence for /usr/src/nbgallery/data (That is where the actual notebooks are stored, the git repo if revisions are enabled, and some other items)

**Comment by mcrutch** [comment]
Also If you could update the readme with info about how your are seeding the SOLR core data (I I think I saw you were using a pvc that had the data in it, but documentation on how that was populated would be useful.)

**Comment by cccs-cat001** [comment]
> I could have missed it (new to helm) but you are going to need some form of persistence for /usr/src/nbgallery/data (That is where the actual notebooks are stored, the git repo if revisions are enabled, and some other items)

That explains the "Notebook empty" error :P

**Comment by cccs-cat001** [comment]
Is there something/someone we're waiting on for this to merge?

**Comment by mcrutch** [comment]
Only waiting on me to have enough time to poke around with this on our cluster.

File: charts/solr/templates/pvc-solr-configs.yaml
Comment (mcrutch) [inline charts/solr/templates/pvc-solr-configs.yaml]: storageClassName should note be tied to specific vendor.  Best to leave empty and allow it to use defaults.

**Comment by mcrutch** [comment]
Now that I finally have time to play with this...I am stuck for the moment because our only cluster is using Helm 2 and I don't have a Helm 3 cluster currently to test with.  I'll probably loop back on this next week after I get around to setting up another cluster, or migrating the cluster to Helm 3 and trying to run JupyterHub on it,

**Comment by cccs-cat001** [comment]
Should probably upgrade to helm 3. Loads of security upgrades...

File: charts/solr/templates/pvc-solr-configs.yaml
Comment (mcrutch) [inline charts/solr/templates/pvc-solr-configs.yaml]: Not compatible with AWS.  Only ReadWriteOnce is supported.

File: values.yaml
Comment (mcrutch) [inline values.yaml]: "default" is actually a bad default.  I think this just has to be left out to pick up the default of the provider (gp2 for AWS for example)

File: templates/deployment.yaml
Comment (mcrutch) [inline templates/deployment.yaml:31]: This should support, at minimum, all of the environment variables available in the default docker-compose, such as oauth

**Review by mcrutch — CHANGES_REQUESTED** [review]

File: values.yaml
Comment (cccs-cat001) [reply values.yaml]: In any kubernetes cluster default can be used. For example, azure's default is azure-disk, but you can use it by just saying `default`

Comment (cccs-cat001) [reply values.yaml]: And this is in the values.yaml file, so the user can easily change the storage class to something they want.

Comment (mcrutch) [reply values.yaml]: IN my testing case, on AWS, EKS, "default" here actually prevented it from picking the cluster default of 'gp2' and thus failed to allocate a volume, which then took a bit of time to figure out what was going on.

File: charts/solr/templates/pvc-solr-configs.yaml
Comment (mcrutch) [reply charts/solr/templates/pvc-solr-configs.yaml]: Really I should say "ReadWriteMany" is only supported by a subset of Kubernetes persistent volume drivers -  (6 out of 19 always with a few others depending on the drive) https://kubernetes.io/docs/concepts/storage/persistent-volumes/

File: values.yaml
Comment (mcrutch) [reply values.yaml:33]: So...had another thought that I might explore which is building a docker container from the official solr docker container, that has the sunspot stuff baked in so it just works.

File: charts/solr/templates/pvc-solr-configs.yaml
Comment (mcrutch) [reply charts/solr/templates/pvc-solr-configs.yaml]: So...had another thought that I might explore which is building a docker container from the official solr docker container, that has the sunspot stuff baked in so it just works.

File: templates/deployment.yaml
Comment (cccs-cat001) [reply templates/deployment.yaml:31]: I think I got them all?

File: values.yaml
Comment (cccs-cat001) [reply values.yaml]: Hmm alright then. Maybe kubernetes isn't as smart as I thought

Comment (cccs-cat001) [reply values.yaml]: would leaving it blank make it take the default?

**Comment by mcrutch** [comment]
Heads up, I just finished this today: https://hub.docker.com/repository/docker/nbgallery/solr
Avoids all of the mapping in of configs and all that fun.  Provides cores for production(default),test and development environments
Will be getting swapped out in the NBGallery docker-compose file when I get to it as well.

**Comment by cccs-cat001** [comment]
Oh that's convenient. I'll use that image instead :)

**Comment by mcrutch** [comment]
For your solr persistence of data, you will want to just have a PVC for /var/solr/data/default/data/ and All of the configmaps go away. (I have been hacking and slashing to run my cluster with the new solr container)

**Comment by mcrutch** [comment]
And can you add 
'''annotations:
  {{ toYaml .Values.service.annotations | indent 4}}'''
to templates/service.yaml

That way if someone tries to run the service directly as a LoadBalancer they can specify the necessary annotations (I am because I am having trouble getting nginx-ingress to work on my cluster right now).

**Comment by cccs-cat001** [comment]
I'll do all this tomorrow/over the weekend :+1:

**Comment by mcrutch** [comment]
Just had an epiphany for the environment variables.  Much like I asked above for the annotations section...how about the bare minimum to get it up is in the deployment.yaml and then you add 
`{{ toYaml .Values.service.env| nindent 12}}` inside the env section (I think 12 is correct) so that you can specify additional environment variables int he values and have them populated.  That way the Facebook/Gitlab/Github/etc.. don't need to be in deployment.yaml and then if someone is using one of those oauth methods (or overwriting any other env var) they can do it through values.yaml.

Thoughts?

**Comment by cccs-cat001** [comment]
I think it's fair to leave the OAuth methods in place, even if the user isn't using those methods. I've seen many helm charts with lots of options that weren't always used. I can still add an "extraEnv" section for any additional variables a user would want to add

**Comment by cccs-cat001** [comment]
And like.. when you deploy without the Oauth stuff enabled, they don't get set as env vars in the deployment (at least I think - did I add {{- if enabled }}?)

**Comment by mcrutch** [comment]
> And like.. when you deploy without the Oauth stuff enabled, they don't get set as env vars in the deployment (at least I think - did I add {{- if enabled }}?)

Yep....and also my wag at the helm syntax was all wrong anyway...lol  do whatever makes sense/works.  Because I want to be able to set the vars like "GALLERY__ANONYMOUS_ACCESS" and "GALLERY__REGISTRATION_ALLOWED" for our deployment.

**Comment by cccs-cat001** [comment]
I can make it possible for you to set extra env vars :)

File: values.yaml
Comment (mcrutch) [inline values.yaml:104]: This does not allow multiple annotations (it throws a json parsing error)  so, for example, to put 
service.beta.kubernetes.io/aws-load-balancer-internal: "true" and 
service.beta.kubernetes.io/aws-load-balancer-type: "nlb"
it fails.

Comment (cccs-cat001) [reply values.yaml:104]: Wow. That's stupid. Because I put the {{ toyaml }} indented it complains. :roll_eyes:

Comment (cccs-cat001) [reply values.yaml:104]: (I'm still fairly new to _creating_ helm charts so it's all trial by error for me too) 
Should be fixed now!

Comment (mcrutch) [reply values.yaml:104]: No worries, you are ahead of me on this..lol

**Comment by mcrutch** [comment]
Ok, one last ask before I merge this...update the README.md since the solr stuff has changed around, but them I am comfortable merging this to master, and if anything else comes up we can handle it by tickets

**Review by mcrutch — APPROVED** [review]

**Pull request merged**
